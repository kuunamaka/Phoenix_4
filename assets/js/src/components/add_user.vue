<!-- adding a new user page -->
<template>
  <table>
    <h2>Adding Users</h2>
    <div>
      <input v-model="name" placeholder="Name">
      <input v-model="username" placeholder="Username">
    </div>
    <button @click="createUser">Create User</button>
    <a href="/users" class="button">Exit</a>
  </table>
</template>

<script>
import axios from 'axios'
export default {
  name: 'add_user',
  data() {
    return {
      name: '',
      username: ''
    }
  },
  methods: {
    async createUser() {
      try {
        const resp = await axios.post('/api/users', {
          user: {
            name: this.name,
            username: this.username
          }
        })
        console.log(resp);
        window.location.href = '/users';
      }
      catch (e) {
        console.log(e)
      }
    }
  }
}
</script>
