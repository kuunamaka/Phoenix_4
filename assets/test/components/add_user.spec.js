import { shallowMount } from "@vue/test-utils";
import _ from "lodash";
import add_user from "@/src/components/add_user.vue";

const axios = require("axios");
const MockAdapter = require("axios-mock-adapter");
const mock = new MockAdapter(axios);

global.window = Object.create(window);
const pathname = "/users/new";

Object.defineProperty(window, "location", {
  value: {
    pathname,
    href: `http://localhost:4001${pathname}`,
  },
});

describe("add_user.vue", () => {
  const wrapper = shallowMount(add_user);
  const vm = wrapper.vm;

  it("renders entire component correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  describe("createUser", () => {
    it("redirect to /users", async () => {
      mock
        .onPost("/api/users", { user: { name: "Test", username: "testtest" } })
        .reply(200);
      await wrapper.setData({ name: "Test", username: "testtest" });

      await vm.createUser();

      expect(window.location.href).toEqual("/users");
    });

    it("show validate error messages", async () => {
      mock
        .onPost("/api/users", { user: { name: "", username: "" } })
        .reply(404);
      await wrapper.setData({ name: "", username: "" });

      const createButton = wrapper.find("button");
      await createButton.trigger("click");

      expect(wrapper.find("p").element).toMatchSnapshot();
    });
  });
});
