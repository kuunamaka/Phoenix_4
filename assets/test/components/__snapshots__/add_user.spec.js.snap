// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add_user.vue is a Vue instance 1`] = `
<table>
  <h2>
    Adding Users
  </h2>
   
  <div>
    <input
      placeholder="Name"
    />
     
    <input
      placeholder="Username"
    />
  </div>
   
  <button>
    Create User
  </button>
   
  <a
    class="button"
    href="/users"
  >
    Exit
  </a>
</table>
`;
